### PR TITLE
upgrade to rails 7.1

### LINKED ISSUE #11673: Upgrade to Rails 7.1
### What we should change and why (this is tech debt)

[The next Rails version has been released.](https://rubyonrails.org/2023/10/5/Rails-7-1-0-has-been-released) There are some performance optimisations which would be nice to have and some new tools we can explore. But there are several steps we need to take first:

- [ ] Wait for all our dependencies to be compatible with Rails 7.1.
- [ ] #7853
- [ ] Address any other existing deprecation warnings.
- [ ] Follow the [upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-7-0-to-rails-7-1).

### Impact and timeline

Not critical at the moment. But good to chip away at this issue.

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- Closes #11673

* [diff file](https://gist.github.com/abdellani/a2c89a905c7092ce89d0b7872c6fbb74)  for `rails app:update`
* upgraded ransack to [v4.1.0](https://github.com/activerecord-hackery/ransack/releases/tag/v4.1.0)
* upgraded paper trail to [v15](https://github.com/paper-trail-gem/paper_trail?tab=readme-ov-file#1a-compatibility)
* in routes declaration, the value for `to:` attribute has to be String  (Symbol is not supported)
* `serialize` requires a `coder`
* require Concerns are no longer needed.
* looks like there's a recursive call on `Spree::Country.cached_find_by`. `cached_find_by` -> `find_by` -> `cached_find_by` ([read more](https://blog.saeloun.com/2022/02/09/rails-prevent-duplicates-in-find_by-cache/))

* libvips is needed to run the project https://stackoverflow.com/questions/70849182/could-not-open-library-vips-42-could-not-open-library-libvips-42-dylib
* `strip_tags` is updated ([check comment](https://github.com/rails/rails-html-sanitizer/issues/124#issuecomment-1140472278))
```ruby
      "I'm friendly!<script>alert('hello! I'm malicious');</script>"
     # will become
      "I'm friendly!alert('hello! I'm malicious');"
    # instead of
      "I'm friendly!"
```
* haml updating to 6.3.0 caused the following error on the tests (increased failures from 5 tests to 23 tests). 

```ruby
    2.2) Failure/Error: - cache(*CacheService::FragmentCaching.ams_shops) do
          
          ActionView::Template::Error:
            undefined method `capture' for "":String
          
                    fragment = output_buffer.capture(&block)
                                            ^^^^^^^^
```
and I faced some issues with connecting the page with Angular components
image

#### What should we test?

- Visit ... page.
- 

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by abdellani** [comment]
I need some help at this level @openfoodfoundation/developers
I'm facing three main issues:
1. `spec/services/order_invoice_generator_spec.rb:52` is passing locally but failing on the CI/CD
2. `spec/models/order_cycle_spec.rb:476` is passing locally but failing on the CI/CD
3. tests related to cache are failing locally and on CI/CD but I was able to run the test manually and I didn't face any issues. it seems to be a problem related to a testing environment

File: Gemfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ gem "image_processing"
 gem 'activemerchant', '>= 1.78.0'
 gem 'angular-rails-templates', '>= 0.3.0'
 gem 'awesome_nested_set'
-gem 'ransack', '~> 2.6.0'
+gem 'ransack', '~> 4.1.0'
```
Comment (mkllnk) [inline Gemfile:20]: I think that we need a separate pull request just for this gem first. There are breaking changes.

* https://github.com/openfoodfoundation/openfoodnetwork/issues/10386

File: Gemfile [hunk 2]
```diff
@@ -93,7 +93,7 @@ gem 'bootsnap', require: false
 gem 'geocoder'
 gem 'gmaps4rails'
 gem 'mimemagic', '> 0.3.5'
-gem 'paper_trail', '~> 12.1'
+gem 'paper_trail', '~> 15.1'
```
Comment (mkllnk) [inline Gemfile:96]: This should also be done in a separate pull request. I don't know the changelog but there's probably a reason why we pinned the version.

File: config/application.rb [hunk 3]
```diff
@@ -223,7 +223,7 @@ module ::Reporting; end
     # Apply framework defaults. New recommended defaults are successively added with each Rails version and
     # include the defaults from previous versions. For more info see:
     # https://guides.rubyonrails.org/configuring.html#results-of-config-load-defaults
-    config.load_defaults 6.1
+    config.load_defaults 7.1
```
Comment (mkllnk) [inline config/application.rb:226]: Maybe let's do this separately as well. We don't have to do it all at once.

File: app/models/spree/country.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@ class Country < ApplicationRecord
 
     validates :name, :iso_name, presence: true
 
-    def self.cached_find_by(attrs)
+    def self.find_by_cached(attrs)
```
Comment (mkllnk) [inline app/models/spree/country.rb:9]: Your commit message says what you are doing here but not why. Generally, I can see the "what" in the diff but the "why" is the important thing to communicate to others. Is there a collision with Rails method naming? If so, should we explain what the difference between our version and Rails' version is?

File: app/controllers/admin/schedules_controller.rb [hunk 5]
```diff
@@ -1,7 +1,6 @@
 # frozen_string_literal: true
 
 require 'open_food_network/permissions'
-require 'order_management/subscriptions/proxy_order_syncer'
```
Comment (mkllnk) [inline app/controllers/admin/schedules_controller.rb:4]: Can we do this in the current Rails version already?

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -59,6 +59,11 @@ jobs:
         with:
           node-version-file: .node-version
 
+      - name: Install libvips
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y libvips
```
Comment (mkllnk) [inline .github/workflows/build.yml:65]: We don't actually use libvips yet. Maybe we should switch to libvips in the current Rails version first.

File: db/schema.rb [hunk 7]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.0].define(version: 20231003000823494) do
+ActiveRecord::Schema[7.1].define(version: 20231003000823494) do
```
Comment (mkllnk) [inline db/schema.rb:13]: This is not related to the sanitizer. Does that  happen when you run db:migrate?

File: spec/models/spree/order_contents_spec.rb [hunk 8]
```diff
@@ -139,6 +139,7 @@
       end
 
       it "updates the order's enterprise fees if completed" do
+        allow(order).to receive(:shipment).and_return(create(:shipment, order:))
```
Comment (mkllnk) [inline spec/models/spree/order_contents_spec.rb:142]: Why do we need this with the new Rails version? May it come from the new defaults you activated in the config?

File: spec/controllers/user_confirmations_controller_spec.rb [hunk 9]
```diff
@@ -55,7 +55,7 @@
         unconfirmed_user.save!
         spree_get :show, confirmation_token: unconfirmed_user.confirmation_token
         expect(response).to be_redirect
-        expect(response.body).to include spree.edit_spree_user_password_path
+        expect(response.headers["Location"]).to include spree.edit_spree_user_password_path
```
Comment (mkllnk) [inline spec/controllers/user_confirmations_controller_spec.rb:58]: What effect does this change have on the UI? Did you test?

**Review by mkllnk — COMMENTED** [review]
Great work!

Rails upgrades tend to become quite big and this pull request is doing a lot of different things. It would be great to split it up. We can create pull requests for the current Rails version, then one minimal pull request for the Rails upgrade and maybe follow-ups to update the config and address deprecations.

And once you split the changes, you may see more easily which cause specs to fail.

**Comment by abdellani** [comment]
Thank you for the review @mkllnk 
I'll split this PR.

File: app/models/spree/country.rb (hunk 4 above)
Comment (abdellani) [reply app/models/spree/country.rb:9]: Sorry, I forgot to explain.

I noticed the collusion after reading this [article](https://blog.saeloun.com/2022/02/09/rails-prevent-duplicates-in-find_by-cache/)
I double checked on Rails repo 
https://github.com/rails/rails/blob/1ea796d6a887ddb6fd7a2395399ae35871aeb00c/activerecord/lib/active_record/core.rb#L246-L257
